### PR TITLE
[MRG+1] Draggable colorbar.

### PR DESCRIPTION
**Pull request by jaeilepp**
Closes https://github.com/mne-tools/mne-python/issues/3005
Also:
closes https://github.com/mne-tools/mne-python/issues/3063

### CODE REVIEW

File: mne/time_frequency/tfr.py [hunk 1]
```diff
@@ -766,7 +767,14 @@ def _onselect(self, eclick, erelease, baseline, mode, layout):
         if 'mag' in self:
             types.append('mag')
         if 'grad' in self:
-            types.append('grad')
+            chs = [ch for ch in self.ch_names if ch.startswith('MEG') and
+                   ch.endswith(('2', '3'))]
```
Comment (agramfort) [inline mne/time_frequency/tfr.py]: this is neuromag specificity. Why is this needed?

Comment (jaeilepp) [inline mne/time_frequency/tfr.py]: Related to https://github.com/mne-tools/mne-python/issues/3063
I'll fix it.

File: mne/viz/utils.py [hunk 2]
```diff
@@ -1107,3 +1107,79 @@ def _compute_scalings(scalings, inst):
         scale_factor = np.max(np.abs(scale_factor))
         scalings[key] = scale_factor
     return scalings
+
+
+class DraggableColorbar(object):
+    """Class for enabling interactive colorbar.
+    See http://www.ster.kuleuven.be/~pieterd/python/html/plotting/interactive_colorbar.html  # doctest: +SKIP
```
Comment (jaeilepp) [inline mne/viz/utils.py]: How can you bypass the flake test on a long line like this?

Comment (Eric89GXL) [inline mne/viz/utils.py]: # noqa

**Comment by kingjr** [comment]
Thanks @jaeilepp for taking the time to do this !

From what I understand this would work for divergent colormaps.

Are you planning on addressing the sequential cases (e.g. combined gradiometers, potentially time frequency spectrum)?  

In matplotlib they typically use one mouse button for displacement and another for scaling, so we could have a similar approach for clim WDYT?

**Comment by jaeilepp** [comment]
Now the color map can be changed with up/down keys. Clicking and dragging changes the scale. Left mouse button moves it up and down and right mouse button changes the lims.
It was pretty straightforward to implement using the ``DraggableColorbar`` in http://www.ster.kuleuven.be/~pieterd/python/html/plotting/interactive_colorbar.html

File: mne/viz/topo.py [hunk 3]
```diff
@@ -256,6 +257,11 @@ def _imshow_tfr(ax, ch_idx, tmin, tmax, vmin, vmax, onselect, ylim=None,
     from matplotlib.widgets import RectangleSelector
 
     extent = (tmin, tmax, freq[0], freq[-1])
+    if cmap == 'interactive':
+        cmap = 'RdBu_r'
+        interactive_cmap = True
+    else:
+        interactive_cmap = False
```
Comment (kingjr) [inline mne/viz/topo.py]:
```
interactive_cmap = False
if cmap == 'interactive':
     cmap = 'RdBu_r'
     interactive_cmap = True
```

**Comment by kingjr** [comment]
LGTM

**Comment by jaeilepp** [comment]
Now interactive mode works for tfr, images and topomaps. For topos it is probably too heavy. Ready for review.

**Comment by Eric89GXL** [comment]
Can you rebase? Then I'll try it and merge if it's all good

File: doc/whats_new.rst [hunk 4]
```diff
@@ -17,6 +17,8 @@ Changelog
 
     - Added the ability to decimate :class:`mne.Evoked` objects with :func:`mne.Evoked.decimate` by `Eric Larson`_
 
+    - Add 'cmap=interactive' option to plotting functions by `Jaakko Leppakangas`_
```
Comment (Eric89GXL) [inline doc/whats_new.rst]: double-backticks on the quote will read nicer, and `interactive` should probably be a string in that code bit

File: mne/evoked.py [hunk 5]
```diff
@@ -476,8 +480,13 @@ def plot_topomap(self, times="auto", ch_type=None, layout=None, vmin=None,
...
         cmap : matplotlib colormap | None
-            Colormap to use. If None, 'Reds' is used for all positive data,
-            otherwise defaults to 'RdBu_r'.
+            Colormap to use. If 'interactive', the colors are adjustable by
+            clicking and dragging the colorbar with left and right mouse
+            button. Left mouse button moves the scale up and down and right
+            mouse button adjusts the range. Up and down arrows can be used to
+            change the colormap. Interactive mode works smoothly only for a
+            small amount of topomaps. If None (default), 'Reds' is used for all
+            positive data, otherwise defaults to 'RdBu_r'.
```
Comment (Eric89GXL) [inline mne/evoked.py]: With this API I can't set the colormap *and* have it interactive, right?

Why not just always have it be interactive?

File: mne/time_frequency/tests/test_tfr.py [hunk 6]
```diff
@@ -375,6 +376,25 @@ def test_plot():
     tfr.plot_topo(picks=[1, 2])
     plt.close('all')
 
+    fig = tfr.plot(picks=[1], cmap='interactive')  # test interactive cmap
+    fig.canvas.key_press_event('up')
+    fig.canvas.key_press_event('down')
+
+    cbar = fig.get_axes()[0].CB  # Fake dragging with mouse.
+    event = backend_bases.MouseEvent('button_press_event', fig.canvas, 0.1,
+                                     0.1, button=1)
```
Comment (Eric89GXL) [inline mne/time_frequency/tests/test_tfr.py]: why not `_fake_click`?

File: mne/time_frequency/tests/test_tfr.py [hunk 7]
```diff
@@ -375,6 +376,25 @@ def test_plot():
...
+    event.inaxes = cbar.cbar.ax
+    cbar.on_press(event)
+    event.y = 0.2
+    cbar.on_motion(event)
```
Comment (Eric89GXL) [inline mne/time_frequency/tests/test_tfr.py]: If it's because you need these, maybe we should add `_fake_*` wrappers for the rest, too. It seems like this should somehow be combined or integrated with the existing `_fake_` code

File: mne/time_frequency/tfr.py [hunk 8]
```diff
@@ -766,13 +774,19 @@ def _onselect(self, eclick, erelease, baseline, mode, layout):
         if 'mag' in self:
             types.append('mag')
         if 'grad' in self:
-            types.append('grad')
+            if (FIFF.FIFFV_COIL_VV_PLANAR_T1 in
+                    np.unique([ch['coil_type'] for ch in self.info['chs']])):
+                chs = [ch for ch in self.ch_names if ch.startswith('MEG') and
+                       ch.endswith(('2', '3'))]
+                if len(chs) < 2:
+                    warn('No grad pairs found.')
```
Comment (Eric89GXL) [inline mne/time_frequency/tfr.py]: `'... found, cannot interactively <do whatever this does>'`

File: mne/viz/evoked.py [hunk 9]
```diff
@@ -71,6 +72,16 @@ def _butterfly_onselect(xmin, xmax, ch_types, evoked, text=None):
...
     ch_types = [type for type in ch_types if type in ('eeg', 'grad', 'mag')]
+    if ('grad' in ch_types and FIFF.FIFFV_COIL_VV_PLANAR_T1 in np.unique(
+            [ch['coil_type'] for ch in evoked.info['chs']])):
+        chs = [ch for ch in evoked.info['ch_names'] if
+               ch.startswith('MEG') and ch.endswith(('2', '3'))]
+        if len(chs) < 2:
+            warn('No grad pairs found.')
+            ch_types.remove('grad')
+            if len(ch_types) == 0:
+                return
```
Comment (Eric89GXL) [inline mne/viz/evoked.py]: This looks like a code dup, can you make it DRY

File: mne/viz/utils.py [hunk 10]
```diff
@@ -1297,3 +1297,80 @@ def _compute_scalings(scalings, inst):
         scale_factor = np.max(np.abs(scale_factor))
         scalings[key] = scale_factor
     return scalings
+
+
+class DraggableColorbar(object):
+    """Class for enabling interactive colorbar.
+    See http://www.ster.kuleuven.be/~pieterd/python/html/plotting/interactive_colorbar.html
```
Comment (Eric89GXL) [inline mne/viz/utils.py]: I don't see any license info there, can you check?

If there is no license, then by default it receives automatically a restrictive license and we cannot include it, so you'd have to contact the author.

**Comment by Eric89GXL** [comment]
Can you mention quickly the examples you've been running to test this to save reviewers some `git grep` calls?

File: mne/evoked.py (hunk 5 above)
Comment (jaeilepp) [inline mne/evoked.py]: If you have many axes (as is the usual case with topomaps for instance) it can be quite laggy. We probably don't want that.

File: mne/viz/utils.py (hunk 10 above)
Comment (jaeilepp) [inline mne/viz/utils.py]: It's under Creative Commons Attribution 3.0 License. See http://www.ster.kuleuven.be/~pieterd/python/html/index.html

**Comment by jaeilepp** [comment]
> Can you mention quickly the examples you've been running to test this to save reviewers some git grep calls?

I've had my own scripts for testing, but it works for tfr, topomap and image. So ``plot_evoked_topomap``, ``plot_channel_epochs_image.py``, ``plot_time_frequency_simulated.py``

File: mne/evoked.py (hunk 5 above)
Comment (Eric89GXL) [inline mne/evoked.py]: That makes sense for my second question about always interactive. But we still have a problem related to my first question about both setting the cmap and making it interactive, right?

**Comment by Eric89GXL** [comment]
Does `plot_evoked_topomap` work for you? It seems *really* slow for me

**Comment by Eric89GXL** [comment]
> For topos it is probably too heavy.

I think we should probably remove the option if it isn't really usable. At the very least we should add a `.. warning::` that it's likely very slow.

**Comment by Eric89GXL** [comment]
Works nicely in `plot_visualize_evoked.py` using a `evoked_r_aud.plot_image(picks=picks, cmap='interactive')` call. I expected the scroll wheel to do the same thing as the right-click-drag, namely "zoom" the colorbar. WDYT?

**Comment by jaeilepp** [comment]
> Does plot_evoked_topomap work for you? It seems really slow for me

Yeah, it's quite slow. I'll investigate.
> I think we should probably remove the option if it isn't really usable. At the very least we should add a .. warning:: that it's likely very slow.

It's not enabled for topographies

**Comment by jaeilepp** [comment]
> Works nicely in plot_visualize_evoked.py using a evoked_r_aud.plot_image(picks=picks, cmap='interactive') call. I expected the scroll wheel to do the same thing as the right-click-drag, namely "zoom" the colorbar. WDYT?

Yeah, why not.

**Comment by Eric89GXL** [comment]
> It's not enabled for topographies

Sorry, I meant for `topomap` plots even, if we can't make them fast enough to actually use.

File: mne/evoked.py (hunk 5 above)
Comment (jaeilepp) [inline mne/evoked.py]: I wonder how the API should be. We probably don't want an extra param just for this. We could accept something like 'Reds interactive'. WDYT?

Comment (agramfort) [inline mne/evoked.py]: can someone give it a try?

Comment (kingjr) [inline mne/evoked.py]: @jaeilepp is it laggy even if we don't actually interact with the figure? Else, I think we're ok having it always interactive, no?

Comment (jaeilepp) [inline mne/evoked.py]: No, but if the user happens to hit the down/up arrow or mouse scroll it stops responding.

Comment (kingjr) [inline mne/evoked.py]: would it be possible to have always `interactive` but topomap until we solve this lagging issue?

Comment (Eric89GXL) [inline mne/evoked.py]: A tuple or dict would probably be better since not all mpl cmaps  (e.g.,
custom ones) are not accessible by string

**Comment by jaeilepp** [comment]
Now cmap is accepted as tuple, where the second element defines the interacitivity. It is also on by default except with topomaps when there are more than two of them.

**Comment by agramfort** [comment]
is it technically possible to make a colorbar interactive after plotting? let's say I double click on the colorbar and it becomes interactive.

**Comment by jaeilepp** [comment]
I think that would be possible. But maybe that is not desirable. You could easily make the topoplots unresponsive by accidentally doubleclicking the colorbar.

**Comment by jaeilepp** [comment]
Looks like there's a painful rebase coming up with https://github.com/mne-tools/mne-python/pull/3275. Maybe that should be merged first. I think I can handle the rebase.

**Comment by jona-sassenhagen** [comment]
I think it can be merged today or tomorrow, right @mmagnuski?

**Comment by agramfort** [comment]
it works for me but API is not super satisfying to me. Currently I have to
remember the name of the cmap I want to use and second it adds True in the
code that the reader of the code has no way of knowing what it means.

can you think of a better API approach?

**Comment by jaeilepp** [comment]
I'm not sure what you mean. It's supposed to work just the way it did before, but with an interactive colorbar. If you want to explicitly set interactivity off, you can pass False as the second argument.

**Comment by agramfort** [comment]
yes it works but I need to do for example

evoked.plot_topomap(times=0.1, cmap=('Reds', True))

so I need to know that cmap is Reds because grads are plotting.

Are I do: cmap=(None, True) and there is no way when reading this
code that True means interactive and None means use the default cmap.

At least it's my feeling.

**Comment by agramfort** [comment]
personally I preferred the API with cmap='interactive'

we use the same API for proj e.g. proj='interactive'

what do others think?

**Comment by mmagnuski** [comment]
@jona-sassenhagen 
Yes, I think I will handle all the comments etc. today/tomorrow so  #3275 can be merged first. Sorry for the pain of rebasing @jaeilepp.

When it comes to API I think I also prefer `cmap='interactive'`, but the ability to set specific `cmap` and turn on interactive cmap mode would still be nice. What about `cmap="Reds", interactive_cmap=True`?

**Comment by Eric89GXL** [comment]
With that API you cannot set the cmap and have it be interactive. But maybe
'interactive' could be a shortcut for Reds True.

**Comment by Eric89GXL** [comment]
(In response to @agramfort)

File: mne/viz/topo.py [hunk 11]
```diff
@@ -249,13 +250,18 @@ def _check_vlim(vlim):
...
                 hline=None):
     """ Aux function to show time-freq map on topo """
     import matplotlib.pyplot as plt
     from matplotlib.widgets import RectangleSelector
 
     extent = (tmin, tmax, freq[0], freq[-1])
+    cmap, interactive_cmap = cmap
+    #if cmap[1]:
+    #    cmap = 'RdBu_r'
+    #    interactive_cmap = True
```
Comment (jaeilepp) [inline mne/viz/topo.py]: Leftovers from changing the API. I'll remove these.

**Comment by jaeilepp** [comment]
I kind of like the solution that's in place now. Now there's basically no API change except that you have an interactive scrollbar without adding any new params. And you also have the freedom to disable it.

**Comment by Eric89GXL** [comment]
@jaeilepp can you make `'interactive'` a shortcut to `('Reds', True)`? That might satisfy @agramfort's use case

**Comment by jaeilepp** [comment]
Yeah, that can be done.

**Comment by Eric89GXL** [comment]
SyntaxError

File: mne/epochs.py [hunk 12]
```diff
@@ -1198,8 +1198,15 @@ def plot_image(self, picks=None, sigma=0., vmin=None,
...
-            Colormap.
+        cmap : matplotlib colormap | (colormap, bool) | 'interactive'
+            Colormap. If tuple, the first value indicates the colormap to use
+            and the second value is a boolean defining interactivity. In
+            interactive mode the colors are adjustable by clicking and dragging
+            the colorbar with left and right mouse button. Left mouse button
+            moves the scale up and down and right mouse button adjusts the
+            range. Hitting space bar resets the scale. Up and down arrows can
+            be used to change the colormap. If 'interactive', translates to
+            ('RdBu_r', True). Defaults to 'RdBu_r'.
```
Comment (agramfort) [inline mne/epochs.py]: AFAIK we don't use RdBu_r all the time eg if we plot grads norm. Write

If 'interactive', the default colormap is used with interactive support.

File: mne/viz/evoked.py [hunk 13]
```diff
@@ -181,7 +188,10 @@ def _plot_evoked(evoked, picks, exclude, unit, show,
                            ' for interactive SSP selection.')
     if isinstance(gfp, string_types) and gfp != 'only':
         raise ValueError('gfp must be boolean or "only". Got %s' % gfp)
-
+    if cmap == 'interactive':
+        cmap = ('RdBu_r', True)
```
Comment (agramfort) [inline mne/viz/evoked.py]: shouldn't it be (None, True) to guess the good default?

**Comment by agramfort** [comment]
besides LGTM

File: mne/epochs.py (hunk 12 above)
Comment (jaeilepp) [inline mne/epochs.py]: This is the epochs image. No grad pairs here.

**Comment by agramfort** [comment]
+1 for merge 

Someone else should test before we merge

**Comment by Eric89GXL** [comment]
From before:

>  I expected the scroll wheel to do the same thing as the right-click-drag, namely "zoom" the colorbar. WDYT?

It looks like instead this was implemented as scroll wheel changes the colormap, is that correct?

**Comment by Eric89GXL** [comment]
It looks like `plot_evoked_topomap` has the new `cmap` option, but not `plot_topomap`, `plot_ica_components`, `plot_tfr_topomap`, `plot_epochs_psd_topomap`, or `plot_psds_topomap`. At least `plot_evoked_topomap`, `plot_ica_components`, and `plot_tfr_topomap` all wrap to `plot_topomap`, so I think you could get all four of those to have interactivity by moving your changes to `plot_topomap` instead of `plot_evoked_topomap`. For the others I think you'd need to modify `plot_psds_topomap` which might not be worth it, depending on how much work it would be.

**Comment by jaeilepp** [comment]
> It looks like instead this was implemented as scroll wheel changes the colormap, is that correct?

Sorry, I somehow misread your comment. I'll change it.

**Comment by jaeilepp** [comment]
> At least plot_evoked_topomap, plot_ica_components, and plot_tfr_topomap all wrap to plot_topomap, so I think you could get all four of those to have interactivity by moving your changes to plot_topomap instead of plot_evoked_topomap.

The interactivity is connected to the colorbar which isn't handled in ``plot_topomap``. I don't think we should break that logic.
But yes, it seems I missed a couple of functions. I'll add it to them.

**Comment by Eric89GXL** [comment]
Good call, I'll set to WIP then

**Comment by jaeilepp** [comment]
Rebased. Ready for review.

**Comment by Eric89GXL** [comment]
LGTM and the interaction is nice in e.g. `plot_channel_epochs_image.py`. +1 for merge from me.

If someone else has you make changes, it might be nice to add `cmap='interactive'` in that example or some other one.

**Comment by dengemann** [comment]
Very nice! Indeed. +1 for showing this in one example or even a gif to be tweeted cc @kingjr

**Comment by jaeilepp** [comment]
> If someone else has you make changes, it might be nice to add cmap='interactive' in that example or some other one.

That's basically the default behavior now, but I'll add it somewhere since I'm adding a test anyway.

**Comment by agramfort** [comment]
@jaeilepp can you just update the viz tutorial(s) to document this?

then +1 for merge

**Comment by agramfort** [comment]
you can add a section of changing the colormap eg in the epochs viz tutorial

**Comment by jaeilepp** [comment]
Ready for review/merge.

**Comment by Eric89GXL** [comment]
Thanks @jaeilepp
